### DISCUSSION TITLE
Slices Pattern + immer + get -> typescript error

### DISCUSSION THREAD

**Discussion by bluescurry** [pmndrs/zustand#2242] [Bug report] [answered]
bearfish.ts
```ts
import { StateCreator } from 'zustand';
import { immer } from 'zustand/middleware/immer'
import { BearActions } from './bear';

export type BearFishActions = {
  removeAllBearFish: () => void;
};


export const createBearFishSlice: StateCreator<
  BearFishActions & BearActions,
  [],
  [["zustand/immer", never]],
  BearFishActions
> = 
  immer((set, get) => ({
    removeAllBearFish: () => {
      // error --->   get: () => BearFishActions
      get().removeAllBears();
    }
  }))

```

get() type should be `BearFishActions & BearActions`，but only get `BearFishActions `.

<img width="886" alt="image" src="https://github.com/pmndrs/zustand/assets/19237129/7fd5ce88-b296-4f1c-8c86-c3b27808bda0">

bear.ts
```ts
import { StateCreator } from 'zustand';
import { immer } from 'zustand/middleware/immer'

export type BearState = {
  bears: number;
};

export type BearActions = {
  addBear: (count?: number) => void;
  removeAllBears: () => void;
};

export const createBearSlice: StateCreator<
  BearState & BearActions,
  [],
  [["zustand/immer", never]],
  BearState & BearActions
> = 
  immer((set) => ({
    bears: 0,
    addBear: (count = 1) => set((state) => {
      state.bears += count;
    }),
    removeAllBears: () => set({ bears: 0 })
  }))

```

**Comment by dbritto-dev** [comment]
@bluescurry since you are using immer for an slice you need to type the immer part too

**Reply by bluescurry** [reply]
I'm sorry, I don't understand what you mean. Could you tell me something more specific, such as how to modify the code to meet typescript conditions, thank you

**Reply by dbritto-dev** [reply]
@bluescurry  sure, just `immer<T>(...)`

> Note: when you type `create<T>()(...)` the types are infered from zustand but in this case since you are using immer middleware outside of zustand create you need to type the immer middleware too

**Reply by bluescurry** [reply]
Could you kindly assist me in documenting the `immer<T>()...`? Specifically, I have been studying for a considerable period of time but haven't found the correct approach to writing generics. Thank you very much

**Reply by dbritto-dev** [reply]
@bluescurry would you mind creating a minimal repro on codesandbox?

**Reply by bluescurry** [reply]
@dbritto-dev https://codesandbox.io/p/sandbox/zustand-h4rsw7?file=%2Fsrc%2Fstore%2Fbearfish.ts%3A6%2C32
The demo I created contains an error on line 17 of `src/store/bearfish.ts`

**Reply by dbritto-dev** [reply, marked as answer]
@bluescurry seems that what you are trying to do is not possible since immer middlware wasn't designed for slices but whole store, is not like you can return a slice, you should be able to get what you want pulling out immer middleware outside you slice creator function. Here you go -> https://codesandbox.io/p/sandbox/zustand-forked-m4xrl6?file=%2Fsrc%2Fstore%2Findex.ts%3A14%2C22

> **Note**: on typescript you must use `create<T>()(...)` the curried version of create not `create(...)`

**Reply by bluescurry** [reply]
I acknowledge your response and appreciate your patience. Thank you for providing me with a clear understanding. I wish you all the best in your future endeavors.

**Comment by bluescurry** [comment]
> @bluescurry would you mind creating a minimal repro on codesandbox?



> @bluescurry seems that what you are trying to do is not possible since immer middlware wasn't designed for slices but whole store, is not like you can return a slice, you should be able to get what you want pulling out immer middleware outside you slice creator function. Here you go -> https://codesandbox.io/p/sandbox/zustand-forked-m4xrl6?file=%2Fsrc%2Fstore%2Findex.ts%3A14%2C22
> 
> > **Note**: on typescript you must use `create<T>()(...)` the curried version of create not `create(...)`
